cmd/snap-confine/ns-support.c

```diff
@@ -306,10 +308,172 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
+		//
+		// TODO: enable this for core distributions. This is complex because on
+		// core the rootfs is mounted in initrd and is _not_ changed (no
+		// pivot_root) and the base snap is again mounted (2nd time) by
+		// systemd. This makes us end up in a situation where the outer base
+		// snap will never match the rootfs inside the mount namespace.
+		bool should_discard =
+		    is_classic ? should_discard_current_ns(base_snap_dev) :
+		    false;
+
+		// Send this back to the parent: 2 - discard, 1 - keep.
+		// Note that we cannot just use 0 and 1 because of the semantics of eventfd(2).
+		debug
+		    ("sending information about the state of the mount namespace (%s)",
+		     should_discard ? "discard" : "keep");
+		if (eventfd_write
+		    (event_fd,
+		     should_discard ? SC_DISCARD_YES : SC_DISCARD_NO) < 0) {
+			die("cannot send information about the state of the mount namespace");
+		}
+		// Exit, we're done.
+		debug
+		    ("support process for mount namespace inspection is about to finish");
+		exit(0);
+	}
+	// This is back in the parent process.
+	//
+	// Enable a sanity timeout in case the read blocks for unbound amount of
+	// time. This will ensure we will not hang around while holding the lock.
+	// Next, read the value written by the child process.
+	sc_enable_sanity_timeout();
+	eventfd_t value = 0;
+	debug("receiving information about the state of the mount namespace");
+	if (eventfd_read(event_fd, &value) < 0) {
+		die("cannot receive information about the state of the mount namespace");
+	}
+	sc_disable_sanity_timeout();
+
+	// Wait for the child process to exit and collect its exit status.
+	errno = 0;
+	int status;
+	if (waitpid(child, &status, 0) < 0) {
+		die("cannot wait for the support process for mount namespace inspection");
+	}
+	if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
+		die("support process for mount namespace inspection exited abnormally");
+	}
+	// If the namespace is up-to-date then we are done.
+	if (value == SC_DISCARD_NO) {
+		debug("the mount namespace is up-to-date and can be reused");
+		return 0;
+	}
+	// The namespace is stale, let's check if we can discard it.
+	debug("the mount namespace is stale and should be discarded");
+	if (sc_cgroup_freezer_occupied(snap_name, (uid_t) - 1)) {
+		// Some processes are still using the namespace so we cannot discard it
+		// as that would fracture the view that the set of processes inside
+		// have on what is mounted.
+		fprintf(stderr,
+			"snap %s cannot use current base snap %s because "
+			"existing process are still using the old revision\n",
+			snap_name, base_snap_name);
+		return 0;
+	}
+	// The namespace is both stale and empty. We can discard it now.
+	debug("discarding stale and empty mount namespace");
+	sc_must_snprintf(mnt_fname, sizeof mnt_fname,
+			 "%s/%s%s", sc_ns_dir, snap_name, SC_NS_MNT_FILE);
+
+	// Use MNT_DETACH as otherwise we get EBUSY.
+	if (umount2(mnt_fname, MNT_DETACH | UMOUNT_NOFOLLOW) < 0) {
```

Comment:
It is interesting the MNT_DETACH is needed here but not in `sc_discard_preserved_ns_group()`. Not a blocker, just interesting...

Reply:
I... honestly don't remember at this hour. Indeed this is interesting. I will investigate it

Reply:
So, I tested this and it does fail without it. I will try to come up with a theory.